### ISSUE TITLE
lint: add unslice checker

### ISSUE THREAD

**Issue by Quasilyte** [PieselBois/kfulint#13]
Find slice expressions that can be simplified to sliced expression.

So, this code will trigger a warning:
```go
s[:] // If s is a slice or string
```
And the alternative we are proposing:
```go
s
```
So, yes, simply `s[:]` => `s`.

In other words: for any 2-index slice expression (as opposed to three-index `x[a:b:c]`) over value `X` of form `X[a:b]` where `a=0` and `b=len(X)` (also includes default index values)
suggest replacing `X[a:b]` with just `X` if type of `X` is a string or slice of any element type.

Real-world example of code that can be detected by this check:
https://github.com/golang/go/blob/adb52cff581b13f06282bd8940a6ab8ee333cb4c/src/strings/replace.go#L450

**Comment by fexolm** [comment]
@ludweeg do you mind, if we do it today?

**Comment by coreyog** [comment]
Can I request that this feature be revisited? I'm running into instances where I'm duplicating a slice so that I can modify the duplicate but retain the original and this check is raising a flag.

For example I'm working with a system with what is functionally 2 databases. For an update, I pull from the main DB, update the objects, write them back to the main DB and reflect those changes in the second DB. If the write to the second DB fails, I want to rollback the first DB to what it was originally. In all my update code I have something along the lines of

    origModels := QueryDatabase(...)
    models := origModels[:] // nolint
    // do stuff with models
    if rollbackRequired {
      UpdateDatabase(origModels)
    }

Super simplified but I hope I got the point across. I feel this is an acceptable use of slicing all elements in a slice but this feature marks it as unnecessary.

**Comment by quasilyte** [comment]
@coreyog, I'm afraid you're not doing what you intended to do.
`xs[:]` is a no-op, it's the same as `xs`.
```go
package foo

// example.go

func a(xs []int) []int {
	return xs
}

func b(xs []int) []int {
	return xs[:]
}
```
Try doing `go tool compile example.go`:
```
"".a STEXT nosplit size=31 args=0x30 locals=0x0
	0x0000 00000 (hello.go:3)	TEXT	"".a(SB), NOSPLIT, $0-48
	0x0000 00000 (hello.go:3)	FUNCDATA	$0, gclocals·74143377807d4351509ac14898b0a46e(SB)
	0x0000 00000 (hello.go:3)	FUNCDATA	$1, gclocals·7d2d5fca80364273fb07d5820a76fef4(SB)
	0x0000 00000 (hello.go:3)	FUNCDATA	$3, gclocals·9fb7f0986f647f17cb53dda1484e0f7a(SB)
	0x0000 00000 (hello.go:4)	PCDATA	$2, $1
	0x0000 00000 (hello.go:4)	PCDATA	$0, $1
	0x0000 00000 (hello.go:4)	MOVQ	"".xs+8(SP), AX
	0x0005 00005 (hello.go:4)	PCDATA	$2, $0
	0x0005 00005 (hello.go:4)	MOVQ	AX, "".~r1+32(SP)
	0x000a 00010 (hello.go:4)	MOVQ	"".xs+16(SP), AX
	0x000f 00015 (hello.go:4)	MOVQ	AX, "".~r1+40(SP)
	0x0014 00020 (hello.go:4)	PCDATA	$0, $2
	0x0014 00020 (hello.go:4)	MOVQ	"".xs+24(SP), AX
	0x0019 00025 (hello.go:4)	MOVQ	AX, "".~r1+48(SP)
	0x001e 00030 (hello.go:4)	RET
	0x0000 48 8b 44 24 08 48 89 44 24 20 48 8b 44 24 10 48  H.D$.H.D$ H.D$.H
	0x0010 89 44 24 28 48 8b 44 24 18 48 89 44 24 30 c3     .D$(H.D$.H.D$0.
"".b STEXT nosplit size=31 args=0x30 locals=0x0
	0x0000 00000 (hello.go:7)	TEXT	"".b(SB), NOSPLIT, $0-48
	0x0000 00000 (hello.go:7)	FUNCDATA	$0, gclocals·74143377807d4351509ac14898b0a46e(SB)
	0x0000 00000 (hello.go:7)	FUNCDATA	$1, gclocals·7d2d5fca80364273fb07d5820a76fef4(SB)
	0x0000 00000 (hello.go:7)	FUNCDATA	$3, gclocals·9fb7f0986f647f17cb53dda1484e0f7a(SB)
	0x0000 00000 (hello.go:8)	PCDATA	$2, $1
	0x0000 00000 (hello.go:8)	PCDATA	$0, $1
	0x0000 00000 (hello.go:8)	MOVQ	"".xs+8(SP), AX
	0x0005 00005 (hello.go:8)	PCDATA	$2, $0
	0x0005 00005 (hello.go:8)	MOVQ	AX, "".~r1+32(SP)
	0x000a 00010 (hello.go:8)	MOVQ	"".xs+16(SP), AX
	0x000f 00015 (hello.go:8)	MOVQ	AX, "".~r1+40(SP)
	0x0014 00020 (hello.go:8)	PCDATA	$0, $2
	0x0014 00020 (hello.go:8)	MOVQ	"".xs+24(SP), AX
	0x0019 00025 (hello.go:8)	MOVQ	AX, "".~r1+48(SP)
	0x001e 00030 (hello.go:8)	RET
	0x0000 48 8b 44 24 08 48 89 44 24 20 48 8b 44 24 10 48  H.D$.H.D$ H.D$.H
	0x0010 89 44 24 28 48 8b 44 24 18 48 89 44 24 30 c3     .D$(H.D$.H.D$0.
```
As you see, both compiled identically.
Linter gives a notice that you're doing something strange.

For further info, please read Go specs or go slices internals.

**Comment by coreyog** [comment]
Wow, I don't remember where I read about `[:]` making a copy but it definitely doesn't behave like I thought it did. Sorry to point the finger.
